File: apps/backend/app/db/models/production.py
"""Journal de production (ADR-0031 §4) — un LOT, pas une pièce.

`trigger` vit ici et nulle part ailleurs : un même déclencheur (« contrôle jeudi sur le chapitre
4 ») engendre un cours, trois fiches, deux quiz et huit cartes. Le poser sur chaque ligne de
contenu, c'est le recopier sur cinq tables et le voir diverger au premier correctif.

**Le modèle anticipe, le code n'anticipe pas.** Toutes les valeurs de `trigger` et de
`authorized_by` sont légales ; seules `manual` et `parent_direct` sont **émises** aujourd'hui. Un
test-verrou interdit les autres tant que leur ADR n'existe pas — patron `content_kind` (six valeurs
au modèle, quatre émises) et `parent_rule` (addendum ADR-0011 §G).
"""

from datetime import datetime

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base

# --- Vocabulaires fermés -----------------------------------------------------------------------

TRIGGERS = ("manual", "request", "agenda", "evidence", "derived", "council")
AUTHORIZED_BY = ("parent_direct", "parent_rule")
RUN_STATUSES = ("queued", "running", "done", "failed")

# Ce qui est réellement ÉMIS. Le reste est modélisé et interdit d'écriture (test-verrou).
#
# `agenda` ouvert le 2026-08-03 (ADR-0035 §1) : un CONTRÔLE, avec un chapitre rattaché et une
# échéance proche, déclenche la production de ce chapitre. `devoir` et `rendu` restent légaux et
# non émis — un devoir reviendrait tous les jours et noierait le régulateur.
#
# ⚠️ `evidence` est écarté EN CONNAISSANCE DE CAUSE, pas par manque de temps : l'agenda est la
# seule source EXOGÈNE du produit, sa légitimité se lit sans modèle (« quelqu'un du monde réel a
# écrit qu'il y avait un contrôle jeudi »). Un déclencheur `evidence` ferait décider ZETIS sur SA
# PROPRE mesure — la boucle se refermerait sur elle-même, et une mesure fausse produirait du
# contenu que rien d'extérieur ne viendrait contredire.
EMITTED_TRIGGERS = ("manual", "agenda")

# `parent_rule` s'émet enfin (ADR-0035 §6) : un lot né du scan satisfait la définition littérale du
# §G.1 — **aucun humain n'a ouvert la pièce, NI CLIQUÉ POUR CE LOT**. Elle est restée légale et non
# émise depuis le 2026-07-28, précisément parce que tout lot partait d'un clic.
EMITTED_AUTHORIZED_BY = ("parent_direct", "parent_rule")

# `trigger` → la FK de référence qu'il DOIT renseigner, et elle seule. `manual` n'en a aucune :
# c'est un geste de Papa, il ne référence rien d'autre que lui-même.
TRIGGER_REFERENCE = {
    "manual": None,
    "request": "content_request_id",
    "agenda": "agenda_item_id",
    "evidence": "skill_id",
    "council": "council_report_id",
    "derived": "skill_id",
}


# --- Vocabulaire du JOURNAL (ADR-0034 §1) ------------------------------------------------------

# Les cinq pièces d'un kit, dans l'ordre où `equip_notion` les produit.
PIECES = ("cours", "fiche", "srs", "quiz", "mindmap")

# --- La demande et le lot ne parlent pas la même langue (ADR-0036 §2) ---------------------------
#
# `content_requests.CONTENT_KINDS` en compte six, `PIECES` cinq — et ce n'est pas un décalage de
# version : ce sont deux vocabulaires nés de deux besoins. Deux écarts, tous deux réels :
#
# - **`card` et `srs` désignent la même chose** sous deux noms (la demande parle comme Massimo,
#   le lot comme les tables) ;
# - **`capsule` n'a aucun producteur** dans l'équipement. Son générateur exige une INSTRUCTION en
#   texte libre — l'intention pédagogique de Papa — qu'une demande `(skill_id, content_kind)` ne
#   porte pas. Constaté au read-before-code du 2026-08-03 ; l'ADR-0036 §3 en a tiré la conséquence :
#   la capsule reste un geste de Papa.
#
# La correspondance est écrite ICI, une fois. Une demande absente de cette table ne se produit pas,
# et l'écran doit le dire plutôt que d'offrir un bouton qui échouerait.
REQUEST_KIND_TO_PIECE = {
    "cours": "cours",
    "fiche": "fiche",
    "mindmap": "mindmap",
    "quiz": "quiz",
    "card": "srs",
}

# `blocked` porte sur la NOTION, pas sur une pièce : le gate du §7 l'a écartée avant tout
# équipement. Sa ligne a `piece = NULL` — une notion silencieusement omise se lirait comme un échec
# de production, alors que c'est un gate qui fonctionne (addendum ADR-0031).
OUTCOMES = ("generated", "skipped", "error", "blocked")


class ProductionRun(Base):
    """Un lot de production : ce qui l'a déclenché, qui l'a autorisé, et où il en est."""

    __tablename__ = "production_runs"
    __table_args__ = (
        # Garde-fou de cohérence, au plus près de la donnée. La contrainte complète
        # (« exactement une FK, cohérente avec `trigger` ») est portée par le service et son
        # test-verrou : l'exprimer entièrement en SQL la rendrait illisible et fragile à chaque
        # nouveau déclencheur. Ici on tient le cas le plus facile à violer par erreur.
        CheckConstraint(
            "(trigger <> 'manual') OR ("
            "agenda_item_id IS NULL AND content_request_id IS NULL "
            "AND council_report_id IS NULL AND skill_id IS NULL)",
            name="ck_production_runs_manual_has_no_reference",
        ),
        # « Exactement un scope » (ADR-0036 §2) — un chapitre, OU une pièce sur une notion.
        # Celle-ci, contrairement à la règle des références ci-dessus, s'exprime entièrement en SQL
        # sans devenir illisible : elle ne dépend d'aucun vocabulaire ouvert. Un lot sans scope
        # produirait dans le vide ; un lot à deux scopes ferait diverger l'exécution de l'affichage.
        CheckConstraint(
            "(chapter_id IS NOT NULL AND scope_skill_id IS NULL AND scope_kind IS NULL) "
            "OR (chapter_id IS NULL AND scope_skill_id IS NOT NULL AND scope_kind IS NOT NULL)",
            name="ck_production_runs_exactly_one_scope",
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("student_profiles.id"), index=True)

    trigger: Mapped[str] = mapped_column(String(20))
    authorized_by: Mapped[str] = mapped_column(String(20))
    status: Mapped[str] = mapped_column(String(15), default="queued", index=True)

    # --- Le SCOPE : sur quoi ce lot a produit ---------------------------------------------------
    # ⚠️ Absent du schéma de l'ADR-0031 §4, et pourtant indispensable : ses colonnes disent
    # POURQUOI on a produit (le déclencheur), jamais SUR QUOI. Un run manuel sur un chapitre
    # n'aurait rien porté de son propre périmètre — donc rien à réafficher, rien à rejouer.
    # v1 : le scope est un chapitre (ADR-0031, « v1 = un chapitre »).
    chapter_id: Mapped[int | None] = mapped_column(ForeignKey("chapters.id"), nullable=True)

    # v2 : le scope peut aussi être UNE PIÈCE sur UNE NOTION (ADR-0036 §2). Une fiche demandée
    # produit une fiche — pas le kit, pas le chapitre.
    #
    # ⚠️ **Deux colonnes neuves, et c'est une décision, pas une facilité.** On aurait pu dériver le
    # scope de `content_request_id`, qui porte déjà la notion et le type. L'ADR-0031 §4 a tranché
    # l'inverse et son motif tient : « ses colonnes disent POURQUOI on a produit, jamais SUR QUOI ».
    # Un lot ne doit pas avoir besoin de son déclencheur pour savoir ce qu'il doit faire.
    #
    # ⚠️ Et **`skill_id` n'est pas réutilisé** : c'est la référence de déclencheur d'`evidence` et
    # `derived` (voir `TRIGGER_REFERENCE`). Une colonne qui vaudrait tantôt « pourquoi » tantôt
    # « sur quoi » serait exactement l'ambiguïté qui a fait rejeter `notion_requests` comme support
    # des demandes de contenu.
    scope_skill_id: Mapped[int | None] = mapped_column(ForeignKey("skills.id"), nullable=True)
    # Une valeur de `PIECES`, jamais de `CONTENT_KINDS` : le lot parle la langue des tables.
    scope_kind: Mapped[str | None] = mapped_column(String(10), nullable=True)

    # --- Les RÉFÉRENCES DE DÉCLENCHEUR : typées, jamais polymorphes ------------------------------
    # Un `trigger_ref_id` générique reproduirait l'ambiguïté qui a fait rejeter `notion_requests`
    # pour les demandes de contenu (« un `skill_id` optionnel qui vaut tantôt inconnu tantôt connu
    # serait ambigu »). Une colonne par origine, nullable, et `TRIGGER_REFERENCE` dit laquelle.
    agenda_item_id: Mapped[int | None] = mapped_column(
        ForeignKey("agenda_items.id"), nullable=True
    )
    content_request_id: Mapped[int | None] = mapped_column(
        ForeignKey("content_requests.id"), nullable=True
    )
    council_report_id: Mapped[int | None] = mapped_column(
        ForeignKey("council_reports.id"), nullable=True
    )
    skill_id: Mapped[int | None] = mapped_column(ForeignKey("skills.id"), nullable=True)

    # --- Avancement : où en est le lot ----------------------------------------------------------
    # Le journal doit savoir où il en est, pas seulement s'il tourne. Deux compteurs plutôt qu'une
    # table d'étapes : l'ORDRE de `scope.plan` est déterministe et testé, donc « les N premières
    # notions sont faites » suffit à reconstituer le détail sans le persister.
    total_notions: Mapped[int | None] = mapped_column(Integer, nullable=True)
    done_notions: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # Où il en est MAINTENANT — ce que `done_notions` ne dit pas (ADR-0034 §2).
    current_skill_id: Mapped[int | None] = mapped_column(
        ForeignKey("skills.id"), nullable=True
    )

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    # ⚠️ `created_at` n'est PAS l'heure de démarrage : le job attend en file (concurrence 1, un
    # seul GPU). Sans cette colonne, la durée d'un lot inclut son attente et ne mesure rien.
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    # Battement de cœur, écrit à chaque notion (il y a déjà un commit par notion : coût nul).
    # Un lot `running` dont le battement a expiré est rendu `stale` PAR LA LECTURE — aucun
    # balayage périodique, aucun ordonnanceur (ADR-0034 §2, doctrine §G.3).
    heartbeat_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class ProductionEvent(Base):
    """Ce que le lot a fait, pièce par pièce (ADR-0034 §1).

    **La donnée existait déjà et partait à la poubelle** : `equip_notion` renvoie
    `generated` / `skipped` / `errors` par pièce, `runner.execute` assemblait le tout dans un
    `results` retourné au job RQ — dont personne ne lit le retour. Seul `done_notions` survivait.
    Cette table retient ce qui était déjà calculé ; elle n'instrumente aucun générateur.

    ⚠️ **Écrite dans la MÊME transaction que l'acte qu'elle trace** — patron `log_learning_event`.
    Un lot interrompu garde le détail de ce qu'il avait fait : le journal d'un crash est
    exactement ce pour quoi on l'écrit.
    """

    __tablename__ = "production_events"
    # Toutes les lectures partent d'un lot et suivent l'ordre de production.
    __table_args__ = (Index("ix_production_events_run_created", "run_id", "created_at"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("production_runs.id"), index=True)
    # Nullable : un lot peut échouer avant d'avoir touché la moindre notion.
    skill_id: Mapped[int | None] = mapped_column(ForeignKey("skills.id"), nullable=True)
    # NULL quand l'événement porte sur la notion entière (`outcome='blocked'`).
    piece: Mapped[str | None] = mapped_column(String(10), nullable=True)
    outcome: Mapped[str] = mapped_column(String(10))
    # Message d'erreur, motif de saut, ou motif de blocage rendu par `select_notions`.
    detail: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))

File: packages/types/src/production.ts
// Couverture de production (page Papa « Couverture ») — contrat de `GET /api/production/*`.
// Décisions : addendum ADR-0011 §E (fraîcheur) et §F (provenance), ADR-0023.

/** État d'une cellule de la matrice, pour UN dérivé d'UNE leçon.
 * `stale` prime sur `validated` : un dérivé périmé EST validé, et c'est précisément
 * l'information que le ✓ ne doit pas masquer. */
export type CellState = "absent" | "pending" | "validated" | "stale" | "blocked";

/** Deux causes de blocage distinctes, parce que l'action à mener diffère :
 * `blocked_lesson` → agir dans Programme ; `blocked_no_course` → rédiger le cours ici. */
export type RowState = "blocked_lesson" | "blocked_no_course" | "ready" | "complete";

/** QUI a laissé passer le contenu (§F) — jamais totalisé, jamais transformé en relance.
 * `null` sur un objet validé = antérieur à la traçabilité (« provenance inconnue »).
 *
 * L'échelle `parent` → `parent_bulk` → `parent_rule` est celle du §G.1 : une décision humaine à
 * granularité d'attention décroissante (la pièce, le lot, la règle). `system` n'est pas dessus —
 * il dit « servi sans relecture PAR DOCTRINE », et reste réservé au quiz.
 *
 * ⚠️ `parent_rule` est **légale et non émise** (ADR-0032) : elle suppose un lot que personne n'a
 * demandé, ce qui n'existe pas encore. Le type l'accepte avant que la donnée la porte. */
export type ValidatedBy = "parent" | "parent_bulk" | "parent_rule" | "system";

export interface CoverageCell {
  state: CellState;
  derived_at: string | null;
  validated_by: ValidatedBy | null;
  /** Cible d'un « Régénérer » (la leçon pour la colonne Cours). `null` si `absent`/`blocked`. */
  object_id: number | null;
}

/** Colonne notion-centrée : une fraction, et **aucun état de fraîcheur** (§E.5). */
export interface CoverageFraction {
  covered: number;
  total: number;
}

/** Une notion de la leçon, et ce qu'elle porte de consommable.
 *  `skill_id` est la cible des générations lancées depuis la matrice. */
export interface CoverageNotionItem {
  skill_id: number;
  name: string;
  has_card: boolean;
  has_capsule: boolean;
}

export interface CoverageNotions {
  cards: CoverageFraction;
  capsules: CoverageFraction;
  items: CoverageNotionItem[];
}

export interface CoverageCells {
  cours: CoverageCell;
  quiz: CoverageCell;
  fiche: CoverageCell;
  mindmap: CoverageCell;
}

/** Les quatre colonnes leçon-centrées, dans l'ordre d'affichage de la matrice. */
export type CoverageCellKey = keyof CoverageCells;

export interface CoverageLesson {
  id: number;
  title: string;
  row_state: RowState;
  cells: CoverageCells;
  notions: CoverageNotions;
}

export interface CoverageChapter {
  id: number;
  title: string;
  lessons: CoverageLesson[];
}

export interface CoverageSubject {
  id: number;
  name: string;
  slug: string;
  chapters: CoverageChapter[];
}

export interface CoverageSchoolYear {
  id: number;
  label: string;
  level: string;
}

export interface CoverageTotals {
  lessons: number;
  lessons_validated: number;
  courses_written: number;
  /** Porte sur quiz · fiche · mindmap UNIQUEMENT — le cours en est la condition, pas un dérivé. */
  derivatives_percent: number;
  pending_count: number;
  stale_count: number;
  orphan_count: number;
}

export interface Coverage {
  school_year: CoverageSchoolYear | null;
  totals: CoverageTotals;
  subjects: CoverageSubject[];
}

export type OrphanType = "fiche" | "mindmap" | "quiz";

export interface ProductionOrphan {
  type: OrphanType;
  id: number;
  title: string;
  subject: string | null;
  archived_at: string | null;
  /** Vrai → suppression désactivée : on n'efface pas l'historique de Massimo pour faire propre. */
  has_history: boolean;
}

// --- Production en lot (ADR-0031) --------------------------------------------------------------

/** Une notion dans l'aperçu d'un lot. `reason` renseigné = bloquée par le gate du §7. */
export interface ProductionNotion {
  skill_id: number;
  name: string;
  reason?: string | null;
}

/** Ce qu'un lot ferait, SANS rien créer. Le gate doit être visible AVANT le clic : sans lui, un
 *  chapitre neuf rendrait « rien produit », que Papa lirait comme un échec. */
export interface ProductionPreview {
  chapter_id: number;
  eligible: ProductionNotion[];
  blocked: ProductionNotion[];
  pending_backlog: number;
  max_pending: number;
}

export type ProductionRunStatus = "queued" | "running" | "done" | "failed";

/** État d'un lot. Un ÉTAT, jamais du contenu. */
export interface ProductionRun {
  id: number;
  status: ProductionRunStatus;
  trigger: string;
  authorized_by: string;
  chapter_id: number | null;
  /** Scope de PIÈCE (ADR-0036 §2) — exclusif de `chapter_id`. `null` = lot de chapitre. */
  scope_skill_id: number | null;
  /** Une valeur de `PIECES` : `cours` | `fiche` | `srs` | `quiz` | `mindmap`.
   *
   *  ⚠️ Le vocabulaire des LOTS, pas celui des demandes : ici c'est `srs`, jamais `card`. */
  scope_kind: string | null;
  /** Nom de la notion visée (jointure serveur) — « une fiche sur la notion 17 » ne se lit pas. */
  scope_skill_name: string | null;
  total_notions: number | null;
  done_notions: number | null;
  /** Avancement RÉEL (0-100), calculé serveur.
   *
   *  ⚠️ **Il compte des NOTIONS, pas des secondes** — donc il ne dit rien d'utile sur un lot-pièce,
   *  qui n'en a qu'une : 0 % pendant toute la durée, puis le lot disparaît. Constaté à l'écran le
   *  2026-08-03. Là où la granularité manque (`total_notions <= 1`), l'affichage doit basculer sur
   *  une estimation — et le dire. Le champ, lui, ne ment pas : il n'a simplement rien à dire. */
  progress_pct: number;
  created_at: string;
  /** Instant de DÉMARRAGE réel — `null` tant que le lot attend son tour.
   *
   *  ⚠️ **C'est lui qui rend l'avancement continu d'une page à l'autre.** Là où le serveur n'a
   *  pas de granularité (lot-pièce), l'écran estime ; une estimation ancrée sur le montage du
   *  composant repart de zéro à chaque navigation, et Papa revoyait « 0 % » sur une production
   *  commencée depuis une minute (constaté le 2026-08-05). Ancrée sur cet instant, elle mesure le
   *  temps écoulé — c'est-à-dire ce qu'elle prétend mesurer. */
  started_at: string | null;
  finished_at: string | null;
  /** 🔴 Durée attendue, **MESURÉE** par le serveur (ADR-0041 §9) — médiane des exécutions
   *  réussies du type de travail que ce lot exécute, × son nombre de notions.
   *
   *  ⚠️ Ajouté le 2026-08-06 pour réparer une régression que la slice C venait d'introduire : les
   *  composants ont cessé d'estimer en dur, or cette vue ne portait pas de quoi estimer — un
   *  lot-PIÈCE retrouvé au retour sur la page perdait sa barre. */
  estimated_ms: number;
}

/** `GET /api/production/runs/active` — le lot en cours, augmenté de **qui l'exécute**.
 *
 *  ⚠️ Ce champ ne vit que sur cette route : la question coûte un aller-retour Redis, et le
 *  Journal aligne des dizaines de lots par page. */
export interface ActiveProductionRun extends ProductionRun {
  /** Un worker écoute-t-il la file ?
   *
   *  ⚠️ **`false` ne veut pas dire « ça va être long », il veut dire « personne ne viendra ».**
   *  Un lot en file sans consommateur n'attend pas son tour : il est arrêté. Le 2026-08-05,
   *  quatre lots ont attendu six heures pendant que l'écran affichait « en file d'attente » — une
   *  vérité littérale qui laissait croire à une file qui avance. */
  worker_alive: boolean;
}

// --- Journal de production et veto (ADR-0034) ---------------------------------------------------

/** ⚠️ `stale` est RENDU par le serveur, jamais stocké : un lot dont le battement de cœur a expiré.
 *  La base ne connaît que quatre statuts, l'écran en montre cinq. */
export type JournalRunStatus = ProductionRunStatus | "stale";

export type PieceKind = "cours" | "fiche" | "mindmap" | "quiz" | "srs";

/** `blocked` porte sur la NOTION (`piece: null`) : le gate du §7 l'a écartée avant production.
 *  Une notion silencieusement omise se lirait comme un échec, alors que c'est un gate qui marche. */
export type EventOutcome = "generated" | "skipped" | "error" | "blocked";

/** Où aller pour débloquer une notion écartée — les trois ids de la convention `pilotageLinks`.
 *
 *  ⚠️ **Résolus SERVEUR.** « Quelle est la leçon de cette notion » a UNE réponse dans le dépôt
 *  (ADR-0037) ; la deviner ici à partir d'un `skill_id` en ferait une quatrième. */
export interface BlockedTarget {
  lesson_id: number;
  chapter_id: number;
  subject_id: number | null;
  /** L'objet PRODUIT, quand la ligne en a produit un — le `?focus=` des pages de pilotage.
   *  `null` sur une ligne bloquée (rien n'a été produit) et sur un cours, qui EST la leçon. */
  object_id: number | null;
}

export interface JournalEvent {
  skill_id: number | null;
  skill_name: string | null;
  piece: PieceKind | null;
  outcome: EventOutcome;
  detail: string | null;
  created_at: string;
  /** `null` s'il n'y a rien à ouvrir : ligne non bloquée, ou notion sans aucune leçon — auquel cas
   *  le motif le dit déjà, et un lien qui mènerait quelque part malgré tout serait pire. */
  target: BlockedTarget | null;
  /** La cause de ce blocage tient-elle **encore** ? `null` hors d'une ligne bloquée.
   *
   *  ⚠️ **Une annotation au PRÉSENT, jamais une correction du passé.** Le motif d'origine reste
   *  exact — il dit ce qui s'est passé — mais il se lit comme un problème actuel une fois la cause
   *  levée. Deux temps, deux phrases : « non produit, cours jamais rédigé · **depuis résolu** ».
   *
   *  ⚠️ `true` = **plus aucun blocage**, pas « le motif d'origine a disparu ». Une notion passée de
   *  « jamais rédigé » à « à valider » reste bloquée : annoncer résolu ferait renoncer Papa au
   *  geste qui reste. */
  resolved: boolean | null;
}

export interface JournalPiece {
  kind: PieceKind;
  id: number;
  label: string;
  /** `null` sur les cartes SRS : aucune étape de validation n'existe pour elles (constat de code,
   *  pas un oubli — c'est pourquoi A0b est verrouillée au palier 3). Jamais « non validé ». */
  validated_by: ValidatedBy | null;
  /** Où ouvrir cette pièce — la liste des pièces est toujours visible, contrairement au détail.
   *  `null` si la leçon n'est pas résoluble (hors année active, chapitre non validé). */
  target: BlockedTarget | null;
  skill_id: number | null;
  skill_name: string | null;
  /** LA question du veto : la consommation ferme la fenêtre, pas l'horloge. */
  consumed: boolean;
}

/** Le régime nommé, tel qu'il a été CAPTURÉ sur le lot.
 *
 *  `"sur_mesure"` = des paliers qui ne composent aucun préréglage — un état légitime, que
 *  `niveau_de` rend déjà. `null` = lot antérieur à la capture : **non enregistré**, jamais
 *  reconstitué depuis les réglages d'aujourd'hui (doctrine §F.4). */
export type JournalZetisMode = "manuel" | "semi" | "autonome" | "sur_mesure";

export interface JournalRun {
  id: number;
  status: JournalRunStatus;
  trigger: string;
  authorized_by: string;
  /** ⚠️ **Le régime de CE lot, pas celui d'aujourd'hui.** C'est lui qui rend le résultat lisible :
   *  un lot qui n'a rien produit sous *Manual* n'est pas une panne, c'est un gate qui a
   *  fonctionné. Sans ce mot, les deux se ressemblent — les lots #21/#22 du 2026-08-04 ont été
   *  lus comme des échecs. */
  zetis_mode: JournalZetisMode | null;
  /** D'où vient `zetis_mode` : `"capture"` (le lot l'a enregistré au démarrage, il fait foi) ou
   *  `"deduit"` (reconstitué de ce que le lot a **fait** — un cours qu'il a rédigé, un dérivé qu'il
   *  a laissé à relire, une origine que seul le régime *Autonome* peut produire). `null` quand rien
   *  ne le prouve.
   *
   *  ⚠️ **Jamais lu des réglages d'aujourd'hui** : ceux-là ont pu changer depuis. La déduction ne
   *  regarde que des actes, qui n'ont pas changé. */
  zetis_mode_source: "capture" | "deduit" | null;
  chapter_id: number | null;
  total_notions: number | null;
  done_notions: number | null;
  current_skill_id: number | null;
  current_skill_name: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  events: JournalEvent[];
  pieces: JournalPiece[];
}

/** ⚠️ Aucun total de provenance, aucun ratio ZETIS/Papa (§F.2) : la provenance est un fait,
 *  jamais un reproche — elle s'affiche par objet et ne se totalise pas. */
/** Un TRAVAIL unitaire au Journal (addendum ADR-0041 §17) — ce qu'il sait, et rien de plus.
 *
 *  ⚠️ **Ni `zetis_mode`, ni `pieces`, ni `events`, donc aucun veto.** Ce n'est pas un oubli : un
 *  `AIJob` ne grave aucun régime d'autonomie et ne tamponne aucune pièce produite. L'écran ne doit
 *  offrir aucun bouton de retrait sur ces lignes — il ne pourrait rien retirer. */
export interface JournalTravail {
  id: number;
  job_type: string;
  /** Le mot que Papa lit (« Cartes de révision · Mitose »), jamais le `job_type`. */
  label: string;
  status: "queued" | "running" | "stale" | "succeeded" | "failed";
  /** L'ORIGINE, pas le régime : hors lot ⇒ `manual` par construction (§3.2). */
  trigger: string;
  skill_id: number | null;
  skill_name: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number | null;
  /** Le motif d'échec, **tel quel** — décision du 2026-08-06, il ne se traduit pas. */
  error: string | null;
}

export interface Journal {
  runs: JournalRun[];
  /** Les travaux unitaires de LA MÊME page (addendum ADR-0041 §16) — à entrelacer par date.
   *
   *  ⚠️ Rendus à part plutôt que mêlés à `runs` : un travail ne porte **ni régime, ni pièces, ni
   *  journal ligne à ligne** (§17), et le glisser dans `JournalRun` l'obligerait à faire semblant.
   *  Les entrelacer côté écran n'est pas un tri côté client : la page est déjà la bonne, découpée
   *  en SQL sur l'union des deux modèles. */
  travaux: JournalTravail[];
  /** Pourquoi les travaux sont absents, quand un filtre les écarte (§18) — `null` sinon.
   *  ⚠️ À AFFICHER : une exclusion muette se lit comme un vide. */
  travaux_exclus: string | null;
  has_more: boolean;
  /** Le nombre de lots RETENUS par le filtre — jamais celui de l'histoire entière.
   *
   *  ⚠️ **Ce n'est pas un compteur de provenance** (§F.2, qui vise les totaux « ZETIS vs Papa »,
   *  lesquels jugeraient) : celui-ci sert la pagination et l'état vide. « 7 sur 23 » est juste ;
   *  « 7 sur 7 » cacherait qu'il existe autre chose, et l'état vide n'aurait plus rien à dire. */
  total: number;
}

/** Ce qu'un retrait emporterait. `removable: false` porte TOUJOURS son motif : un refus muet se
 *  lit comme une panne. */
export interface VetoPreview {
  removable: boolean;
  reason: string | null;
  /** Ce que le retrait d'un COURS emporte — il est la source canonique de ses dérivés. */
  cascade: Partial<Record<PieceKind, number[]>>;
}

export interface VetoRemoval {
  removed: Partial<Record<PieceKind, number>>;
}

// --- L'activité de production (ADR-0041) -------------------------------------------------------
//
// La source UNIQUE de toutes les barres de progression Papa : le header et les pages lisent le
// même endpoint. Ce qui disparaît avec ce type, ce sont les 23 constantes de durée en dur — la
// rédaction d'un cours en portait CINQ différentes selon l'écran d'où on la lançait.

/** Un travail en cours, en file, ou échoué — lot pédagogique ou travail unitaire. */
export interface ActivityItem {
  /** `run` = un lot (`ProductionRun`) ; `job` = un travail unitaire (`AIJob`). */
  kind: "run" | "job";
  id: number;
  label: string;
  status: "queued" | "running" | "stale" | "failed";
  /** ⚠️ `null` = **indéterminé**, et JAMAIS `0` pour dire « ça démarre ». Zéro n'est pas une
   *  valeur basse, c'est une absence de mesure — le 2026-08-05, quatre lots arrêtés affichaient
   *  0 %, lu comme « ça commence ». */
  pct: number | null;
  /** Deux régimes de vérité (§6) : `true` = progression RÉELLE calculée serveur (« 7 / 19 ») ;
   *  `false` = estimation ancrée sur `started_at` (« ≈ 40 % »). Un appel LLM n'a aucun grain
   *  interne : les confondre uniformiserait un mensonge. */
  pct_is_measured: boolean;
  /** La FRACTION en **pièces** — « 7 / 19 » (addendum 2 §20). C'est elle qui prouve la mesure :
   *  « 37 % » seul ne se distingue pas d'une estimation bien tournée.
   *
   *  ⚠️ `null` **avec** `pct` et sous exactement la même condition serveur — il existe une fenêtre
   *  où un lot est `running` sans ses compteurs. Les traiter séparément ferait afficher
   *  `null / null · 37 %`. Un travail unitaire n'a aucune pièce : toujours `null`. */
  pieces_done: number | null;
  pieces_total: number | null;
  /** Ce qui est **vraiment** tombé dans la boîte — les pièces `generated` seules, jamais les
   *  `skipped` qui y étaient déjà. Toujours servi (un `COUNT` est exact même hors régime mesuré) :
   *  c'est le badge du stock, pas l'avancement. */
  pieces_produced: number;
  /** La pièce en cours dans la notion en vol — `cours` · `fiche` · `srs` · `quiz` · `mindmap`,
   *  `null` entre deux notions et à la fin.
   *
   *  🔴 **C'est ce champ qui fait bouger la barre.** Les cinq lignes de journal d'une notion
   *  atterrissent d'un seul coup à sa fin : sans cette position, un compte de pièces avancerait
   *  exactement comme un compte de notions (`5/155` = `1/31`), toutes les ~69 s. Un changement de
   *  valeur ici veut dire qu'une pièce vient d'être finie — c'est ce qui lance un jeton sur le
   *  tapis, avec son vrai nom. */
  current_piece: string | null;
  started_at: string | null;
  /** DÉRIVÉ, jamais stocké (§3.2) : un lot porte son déclencheur, un travail hors lot est
   *  manuel par construction. */
  trigger: string | null;
  error: string | null;
  /** 🔴 La durée attendue, **MESURÉE par le serveur** (§9) — médiane des dernières exécutions
   *  réussies de ce type de travail, amorce tant qu'il n'y a pas d'histoire.
   *
   *  C'est ce champ qui a tué les vingt-trois constantes de durée des composants Papa : une barre
   *  locale ne devine plus, elle lit. Deux surfaces ne peuvent plus annoncer deux nombres pour le
   *  même travail, puisqu'il n'y a plus qu'un nombre. */
  estimated_ms: number;
}

export interface ProductionActivity {
  /** Ce qui tourne ; à défaut, le premier de la file. */
  current: ActivityItem | null;
  /** Profondeur de file — jamais un arriéré (§7) : il retombe à zéro tout seul. */
  queued_count: number;
  /** Ce qui attend derrière, **dans l'ordre où la file sera servie** (§7).
   *
   *  ⚠️ Borné à 20 par le serveur, alors que `queued_count` dit le total : une troncature qui ne
   *  se déclare pas se lit comme une exhaustivité. Comparer les deux, c'est savoir s'il en reste. */
  queued: ActivityItem[];
  /** Les échecs NON acquittés. Ils restent jusqu'au clic, pas six secondes. */
  failed: ActivityItem[];
  /** ⚠️ `null` = « la question n'a pas été posée », ce qui n'est PAS `false`. Tester `=== false`. */
  worker_alive: boolean | null;
}
